### PR TITLE
gen4: Fast aggregations

### PR DESCRIPTION
**Pull request by vmg**
## Description

Yay! Really excited for my first project after returning from paternity leave. This has been in my TODO list for many months now, I spotted it while doing code review earlier this year.

Here's the gist of it: the gen4 query execution engine has a performance problem with aggregations. For one single reason: **because it uses serialized SQL values to hold its aggregation context**.

The execution engine was implemented like that because that's what APIs provided by the `evalengine` allowed. All the "exposed API" of the evaluation engine, such as `evalengine.NullSafeCompare`, `evalengine.NullSafeAdd`, `evalengine.Min`, `evalengine.Max`, etc, is implemented the same way. The functions take two `sqltypes.Value` as inputs and return a `sqltypes.Value`. But this is not what we want to be doing for aggregations! Calling these APIs in a loop is incredibly wasteful.

Let me give you some examples of the way aggregations are currently implemented:

- `COUNT(x)`: we have a `sqltypes.Value` as the current count. To increase it, we call `NullSafeAdd` with the current count and a **constant** `Value` which is simply `INT64(1)`. The evalengine parses the current count, parses the `1`, performs an addition, then serializes the new count as a `Value` and returns it. Repeat for every single row in the `Result` set for the query.
- `SUM(x)`: again, we have a `sqltypes.Value`. To sum rows into it, we call `NullSafeAdd` with the current accumulated sum and the value for the new row. Parse the accumulated sum, parse the row, add them together, serialize again. Repeat over for each run.
- `MIN(x)`/`MAX(x)`: the current min/max is in a `Value`. So, call `NullSafeCompare`, which parses the current min/max, parses the new row, compares them and returns the new min/max in its serialized form.
- `CONCAT(x)`: the current concatenation is a `Value`, which cannot be appended to. So for every row that must be concatenated: parse the current `Value`, parse the `Value` for the new row, join them together, create a new `Value`, replace the existing aggregation with the new `Value`.
- `GTID_AGG(x)`: this one's my favourite. Vitess implements GTID aggregation, which merges GTID rows into an aggregated ProtoBuf-based data structure. Obviously aggregation _must_ happen using `sqltypes.Value`. So for every row: de-serialize the current aggregation value, which is stored as a serialized ProtoBuf in a `VARBINARY` inside of a `Value`. Modify the ProtoBuf object. Serialize it again into a brand new `VARBINARY`. Replace the existing one. At the end of the aggregation, de-serialize the `VARBINARY` one last time and serialize it into its textual form in a `VARCHAR`. That's the result of the aggregation!

Alright, that's enough context. We can see that we're missing APIs in the `evalengine` which forces us to implement an _extremely inefficient form of SQL aggregations_. This PR implements optimizes APIs using some tricks I gathered while working on Boost.

The idea is as follows: the aggregation state is not kept in `sqltypes.Value`, but in specialized "aggregation states" which implement the `aggregator` interface in the `engine` package. These aggregators are specialized for every aggregation type: they are **incrementally updated in-place** with the incoming rows of the aggregation. They only serialize the `sqltypes.Value` _once_, at the end of the aggregation.

Here are some examples of how these work:

- `COUNT(x)`: this aggregator has an `int64` inside of it. When a row comes in, we add 1 to the int64, using the power of the ADDL instruction of our CPU. The `evalengine` is not involved.
- `SUM(x)` this aggregator is specialized based on the type of the column being aggregated. We have specializations for signed, unsigned, float and decimal columns. Each specialization holds an internal sum with the native datatype (e.g. `int64` for signed, `float64` for floats, etc). When updated, we parse the `sqltypes.Value` of the incoming row and perform the addition using, huh, a single CPU instruction (a few more for `DECIMAL`).
    - We use a neat trick for signed and unsigned columns: note that although these are integral datatypes, MySQL aggregates them as `DECIMAL`. What we do now in Vitess is aggregating them in a native type (`int64` or `uint64`) _until_ the type overflows, and only then we start using our `Decimal` implementation to keep count. _Neat_.
- `MIN(X)`, `MAX(x)`: just like `SUM`, these are specialized for the different numeric types. Much more efficient to compare two `int64` than to compare two `sqltype.Value`s with `NullSafeComparison`. There's a generic fallback for all the non-numeric types that must be compared "the slow way".
- `CONCAT(x)`: this aggregator contains a `[]byte` slice and appends into it. Heh. It's so easy to implement these things when you don't have to serialize them from/into `sqltypes.Value`.
- `GTID_AGG(x)`: this aggregator contains the ProtoBuf object with the GTIDs and it just updates them incrementally. It never de-serializes it. It serializes it once into its textual form at the end of the aggregation. :sparkles: 

So, that's the basic idea. Change the APIs, and suddenly aggregation becomes fast, because the intermediate state is kept as native Go data types instead of serialized from/to `sqltypes.Value` over and over again. How much faster? Glad you asked:

```
goos: linux
goarch: amd64
pkg: vitess.io/vitess/go/vt/vtgate/engine
cpu: 13th Gen Intel(R) Core(TM) i9-13900K
                                                                            │    old.txt     │               new.txt               │
                                                                            │     sec/op     │   sec/op     vs base                │
ScalarAggregate/count(INT64)/TryExecute-9                                      1227.56µ ± 0%   33.45µ ± 0%  -97.28% (p=0.000 n=10)
ScalarAggregate/sum_small(INT64)/TryExecute-9                                   2761.2µ ± 1%   152.4µ ± 0%  -94.48% (p=0.000 n=10)
ScalarAggregate/sum(INT64)/TryExecute-9                                         4389.8µ ± 0%   682.4µ ± 0%  -84.45% (p=0.000 n=10)
ScalarAggregate/sum(FLOAT64)/TryExecute-9                                       3298.4µ ± 0%   616.5µ ± 2%  -81.31% (p=0.000 n=10)
ScalarAggregate/sum(DECIMAL)/TryExecute-9                                        4.420m ± 0%   1.204m ± 0%  -72.76% (p=0.000 n=10)
ScalarAggregate/min(INT64)/TryExecute-9                                          840.3µ ± 1%   310.9µ ± 1%  -63.00% (p=0.000 n=10)
ScalarAggregate/min(FLOAT64)/TryExecute-9                                       1779.1µ ± 1%   599.8µ ± 1%  -66.29% (p=0.000 n=10)
ScalarAggregate/min(DECIMAL)/TryExecute-9                                       2358.2µ ± 1%   861.4µ ± 0%  -63.47% (p=0.000 n=10)
ScalarAggregate/min(VARBINARY)/TryExecute-9                                      408.6µ ± 1%   248.2µ ± 0%  -39.25% (p=0.000 n=10)
ScalarAggregate/keyspace(VARCHAR)_gtid(VARCHAR)_shard(VARCHAR)/TryExecute-9   4338.310m ± 2%   6.242m ± 1%  -99.86% (p=0.000 n=10)
geomean                                                                          4.104m        479.5µ       -88.32%

                                                                            │     old.txt     │               new.txt                │
                                                                            │      B/op       │     B/op      vs base                │
ScalarAggregate/count(INT64)/TryExecute-9                                       960734.0 ± 0%     568.0 ± 0%  -99.94% (p=0.000 n=10)
ScalarAggregate/sum_small(INT64)/TryExecute-9                                  3184609.5 ± 0%     600.0 ± 0%  -99.98% (p=0.000 n=10)
ScalarAggregate/sum(INT64)/TryExecute-9                                         4515.3Ki ± 0%   938.2Ki ± 0%  -79.22% (p=0.000 n=10)
ScalarAggregate/sum(FLOAT64)/TryExecute-9                                      1411136.0 ± 0%     632.0 ± 0%  -99.96% (p=0.000 n=10)
ScalarAggregate/sum(DECIMAL)/TryExecute-9                                        4.658Mi ± 0%   1.688Mi ± 0%  -63.76% (p=0.000 n=10)
ScalarAggregate/min(INT64)/TryExecute-9                                         640683.0 ± 0%     536.0 ± 0%  -99.92% (p=0.000 n=10)
ScalarAggregate/min(FLOAT64)/TryExecute-9                                       800704.5 ± 0%     560.0 ± 0%  -99.93% (p=0.000 n=10)
ScalarAggregate/min(DECIMAL)/TryExecute-9                                       2802.7Ki ± 0%   896.6Ki ± 0%  -68.01% (p=0.000 n=10)
ScalarAggregate/min(VARBINARY)/TryExecute-9                                     640668.0 ± 0%     552.0 ± 0%  -99.91% (p=0.000 n=10)
ScalarAggregate/keyspace(VARCHAR)_gtid(VARCHAR)_shard(VARCHAR)/TryExecute-9   12653.00Mi ± 0%   11.08Mi ± 0%  -99.91% (p=0.000 n=10)
geomean                                                                          3.873Mi        14.82Ki       -99.63%

                                                                            │    old.txt     │               new.txt               │
                                                                            │   allocs/op    │  allocs/op   vs base                │
ScalarAggregate/count(INT64)/TryExecute-9                                      50005.00 ± 0%    11.00 ± 0%  -99.98% (p=0.000 n=10)
ScalarAggregate/sum_small(INT64)/TryExecute-9                                 119990.00 ± 0%    12.00 ± 0%  -99.99% (p=0.000 n=10)
ScalarAggregate/sum(INT64)/TryExecute-9                                         139.99k ± 0%   20.01k ± 0%  -85.71% (p=0.000 n=10)
ScalarAggregate/sum(FLOAT64)/TryExecute-9                                      69076.00 ± 0%    14.00 ± 0%  -99.98% (p=0.000 n=10)
ScalarAggregate/sum(DECIMAL)/TryExecute-9                                       132.83k ± 0%   43.10k ± 0%  -67.55% (p=0.000 n=10)
ScalarAggregate/min(INT64)/TryExecute-9                                        10009.00 ± 0%    12.00 ± 0%  -99.88% (p=0.000 n=10)
ScalarAggregate/min(FLOAT64)/TryExecute-9                                      30007.00 ± 0%    14.00 ± 0%  -99.95% (p=0.000 n=10)
ScalarAggregate/min(DECIMAL)/TryExecute-9                                        75.12k ± 0%   24.05k ± 0%  -67.98% (p=0.000 n=10)
ScalarAggregate/min(VARBINARY)/TryExecute-9                                    10009.00 ± 0%    11.00 ± 0%  -99.89% (p=0.000 n=10)
ScalarAggregate/keyspace(VARCHAR)_gtid(VARCHAR)_shard(VARCHAR)/TryExecute-9   200308.1k ± 0%   110.1k ± 0%  -99.95% (p=0.000 n=10)
geomean                                                                          113.3k         308.5       -99.73%
```

Geometric mean between all benchmarks: -88.32% reduction in CPU runtime, -99.63% reduction in allocation size (from 3.8mb allocated per benchmark run to 14.82kb), -99.73% reduction in allocation count (from 113300 allocations per benchmark run to 308).

I can show you a graph for CPU:

I didn't render Memory graphs because they're too skewed. The GTID aggregation went from 200mb allocated per run to 100kb, which skews too much the visualization even in a logarithmic scale. :sleepy: 

cc @systay @frouioui @dbussink 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/evalengine/api_aggregation.go [hunk 1]
```diff
@@ -0,0 +1,481 @@
+package evalengine
```
Comment (systay) [inline go/vt/vtgate/evalengine/api_aggregation.go:1]: license missing

**Comment by dbussink** [comment]
@vmg There's a number of test failures here that seem related?

```
=== RUN   TestBuggyQueries
    mysql.go:219: Query (select /*vt+ PLANNER=Gen4 */ min(tbl0.loc) from dept as tbl0) results mismatched.
        Vitess Results:
        [VARCHAR("CHICAGO")]
        MySQL Results:
        [VARCHAR("BOSTON")]
        query plan: 
        {
        	"OperatorType": "Aggregate",
        	"Variant": "Scalar",
        	"Aggregates": "min(0 COLLATE utf8mb4_general_ci) AS min(tbl0.loc)",
        	"Inputs": [
        		{
        			"OperatorType": "Route",
        			"Variant": "Scatter",
        			"Keyspace": {
        				"Name": "ks_random",
        				"Sharded": true
        			},
        			"FieldQuery": "select min(tbl0.loc) from dept as tbl0 where 1 != 1",
        			"Query": "select /*vt+ PLANNER=Gen4 */ min(tbl0.loc) from dept as tbl0",
        			"Table": "dept"
        		}
        	]
        }
  
```

They are all for aggregate functions so I think there's an issue here somewhere.

**Comment by harshit-gangal** [comment]
error looks legit

```
2023-08-31T15:48:53.0284328Z         Vitess Results:
2023-08-31T15:48:53.0284535Z         [VARCHAR("ACCOUNTING") VARCHAR("cattle")]
2023-08-31T15:48:53.0284656Z         MySQL Results:
2023-08-31T15:48:53.0284865Z         [VARCHAR("SALES") VARCHAR("cattle")]
```

**Comment by vmg** [comment]
Yeah I did an oopsie with a `<` sign. Less than... Greater than... does it really matter that much?

File: go/vt/vtgate/semantics/typer.go [hunk 2]
```diff
@@ -55,15 +55,15 @@ func (t *typer) up(cursor *sqlparser.Cursor) error {
...
 			t, ok := t.exprTypes[arg]
 			if ok {
-				inputType = &t.Type
+				inputType = t.Type
 			}
 		}
-		typ, _ := code.Type(inputType)
-		t.exprTypes[node] = Type{Type: typ, Collation: collations.DefaultCollationForType(typ)}
+		type_ := code.Type(inputType)
+		t.exprTypes[node] = Type{Type: type_, Collation: collations.DefaultCollationForType(type_)}
```
Comment (dbussink) [inline go/vt/vtgate/semantics/typer.go:66]: I liked `typ`, especially since @systay told me it's Swedish for `type` :wink:.

**Review by dbussink — APPROVED** [review]

**Pull request merged**